src/descriptor/mod.rs

```diff
@@ -771,6 +804,46 @@ impl Descriptor<DefiniteDescriptorKey> {
         let derived = self.translate_pk(&mut Derivator(secp))?;
         Ok(derived)
     }
+
+    /// Returns the set of keys which are available in the [`KeyMap`] in form of [`Assets`]
+    pub fn available_keys(&self, key_map: &KeyMap) -> Assets {
+        let mut available_keys = vec![];
+
+        self.for_each_key(|pk| {
+            let found = match &pk.0 {
+                s @ DescriptorPublicKey::Single(_) => key_map.contains_key(&s),
+                DescriptorPublicKey::XPub(xkey) => {
+                    if key_map.contains_key(&DescriptorPublicKey::XPub(xkey.clone())) {
+                        true
+                    } else if xkey.derivation_path.len() > 0 {
+                        let unwind_wildcard = DescriptorXKey {
+                            origin: xkey.origin.clone(),
+                            xkey: xkey.xkey,
+                            wildcard: Wildcard::Unhardened,
+                            derivation_path: xkey
+                                .derivation_path
+                                .into_iter()
+                                .take(xkey.derivation_path.len() - 1)
+                                .cloned()
+                                .collect::<Vec<_>>()
+                                .into(),
+                        };
```

Comment:
In dc0b2804a8368bea9f5e6e4717ad31bc8933e090:

This "remove the wildcard" business should be a method on `DescriptorPublicKey`, if this is the correct logic ... though TBH I'm not certain that it is. Where does `key_map` come from and why would it have wildcard-stripped xpubs in it?

Reply:
Sure, I can definitely move this in a method on `DescriptorPublicKey`. Tomorrow I'll look at the logic more closely and maybe add some tests for it, but the rationale is:

- `Assets` works with `DefiniteDescriptorKey`s, because sometimes we have to refer to them given the hash (so we can't try bruteforcing derivation indexes on `DescriptorPublicKey`).
- When we encounter an "XPub" variant of a `DefiniteDescriptorKey` it means we have the derived version of a `DescriptorPublicKey`, and there are two possible cases there:
    - If the `DescriptorPublicKey` had a wildcard, then in the definite key we will find a path that has the exact same prefix plus the last element which has whatever derivation index the key have been derived to. So this is why when the derivation index is not empty we try looking up a key that doesn't have the last index in the derivation path plus the wildcard
    - If the `DescriptorPublicKey` didn't have a wildcard at all, then deriving it doesn't change it in any way, which is why we also try looking up the key directly, in case it was unchanged.

I think this, at least on paper, makes sense but I'll double check this tomorrow.

The `key_map` comes from the descriptor parsing, when you parse a descriptor with secrets you get back the `Descriptor<DescriptorPublicKey>` plus the map public -> secret keys of all the secret keys found in the descriptor. Normally the user would then derive the descriptor using `at_derivation_index()`, and then call this method on the derived descriptor, giving the original key map as an argument so that we can figure out which keys you actually know the secret for.

Ideally we would just be working on `Descriptor<DescriptorPublicKey>`, to essentially get two benefits:
1. The `Assets` structure would not be tied to the derived descriptor, but it could be created once and reused for every derivation index
2. We would not have to do this weird "unwinding" of the wildcard

Unfortunately as I was saying above, sometimes we have to answer lookups that only give us the key hash, so we can't work with keys that have wildcards because we would have to bruteforce them. For example the `lookup_raw_pkh_pk(&self, hash: &hash160::Hash)` method would be very hard to implement if the assets only contained underived keys.

Reply:
My feeling is that we should make `Assets` hold `DescriptorPublicKey`s and we should implement `lookup_raw_pkh_pk` on a best-effort basis. Meaning, we store hashes for `DescriptorPublicKey`s without wildcards and don't for ones with wildcards.

I'm comfortable doing this because since #431 we've treated raw pkhs as second-class-citizens, which can only occur when you are inferring descriptors directly from `Script`. If you have a stringified descriptor then you will always have actual public keys and not need to do lookups by hash. This is also the only form of pkh that Core supports, BTW.